File: kiwi-admin/backend/src/main/java/com/kiwi/project/bpm/model/BpmProcessIoInventory.java
package com.kiwi.project.bpm.model;

import lombok.Data;

import java.util.ArrayList;
import java.util.List;

/**
 * 按组件拓扑序展开的流程 IO 清单（设计时配置，不含运行时实例变量）。
 */
@Data
public class BpmProcessIoInventory {

    private List<Node> nodes = new ArrayList<>();
    private List<StartVariable> startVariables = new ArrayList<>();
    /**
     * 全流程目录输出并集，按拓扑序合并，同名 key 以后出现的组件为准。
     * 供另存为组件等流程级契约使用。
     */
    private List<BpmComponentParameter> processOutputs = new ArrayList<>();

    @Data
    public static class Node {
        private String nodeId;
        private String name;
        private String componentId;
        private String componentName;
        private List<Param> inputs = new ArrayList<>();
        private List<Param> outputs = new ArrayList<>();
    }

    @Data
    public static class Param {
        private String key;
        private String name;
        private String description;
        private boolean required;
        private Direction direction;
        private boolean filled;
        private ValueKind valueKind;
        private String configuredValue;
        private List<String> expressionRefs = new ArrayList<>();
        private boolean satisfiedByUpstream;
        /** 输出写入的流程变量名（目录 defaultValue 或 key）。 */
        private String processVariable;
    }

    @Data
    public static class StartVariable {
        private String key;
        private String name;
        private String description;
        private boolean required;
        private String nodeId;
        private String parameterKey;
    }

    public enum Direction {
        Input,
        Output
    }

    public enum ValueKind {
        Empty,
        Literal,
        Expression
    }
}

File: kiwi-admin/frontend/src/app/pages/bpm/types/bpm-process-io-inventory.ts
/** 与后端 {@code BpmProcessIoInventory} 对齐（Jackson 枚举名为 PascalCase）。 */
export type BpmProcessIoValueKind = 'Empty' | 'Literal' | 'Expression';
export type BpmProcessIoDirection = 'Input' | 'Output';

export interface BpmProcessIoParam {
  key: string;
  name?: string;
  required?: boolean;
  direction?: BpmProcessIoDirection;
  filled?: boolean;
  valueKind?: BpmProcessIoValueKind;
  configuredValue?: string | null;
  expressionRefs?: string[];
  satisfiedByUpstream?: boolean;
  processVariable?: string;
}

export interface BpmProcessIoNode {
  nodeId: string;
  name?: string;
  componentId?: string;
  componentName?: string;
  inputs?: BpmProcessIoParam[];
  outputs?: BpmProcessIoParam[];
}

export interface BpmProcessIoStartVariable {
  key: string;
  name?: string;
  description?: string;
  required?: boolean;
  nodeId?: string;
  parameterKey?: string;
}

export interface BpmProcessIoInventory {
  nodes?: BpmProcessIoNode[];
  startVariables?: BpmProcessIoStartVariable[];
  processOutputs?: unknown[];
}
